static/app/utils/url/useLocationQuery.tsx

```diff
@@ -10,30 +10,46 @@ type Decoder = typeof decodeList | typeof decodeScalar | typeof decodeInteger;
  * Select and memoize query params from location.
  * This returns a new object only when one of the specified query fields is
  * updated. The object will remain stable otherwise, avoiding re-renders.
+ *
+ * You shouldn't need to manually set the `InferredRequestShape` or `InferredResponseShape`
+ * generics, instead type the left side of the statement.
+ *
+ * For example:
+ * ```
+ * type QueryFields = {statsPeriod: string};
+ * const query: QueryFields = useLocationQuery({
+ *   fields: {statsPeriod: decodeScalar}
+ * });
+ * ```
  */
 export default function useLocationQuery<
-  Fields extends Record<string, Scalar | Scalar[]>,
->({fields}: {fields: Record<keyof Fields, Fields[string] | Decoder>}): Fields {
+  InferredRequestShape extends Record<string, Scalar | Scalar[] | Decoder>,
+  InferredResponseShape extends {
+    readonly [Property in keyof InferredRequestShape]: InferredRequestShape[Property] extends Decoder
+      ? ReturnType<InferredRequestShape[Property]>
+      : InferredRequestShape[Property];
+  },
+>({fields}: {fields: InferredRequestShape}): InferredResponseShape {
   const location = useLocation();
 
   const locationFields = {};
-  const staticFields = {};
+  const forwardedFields = {};
   Object.entries(fields).forEach(([field, decoderOrValue]) => {
     if (typeof decoderOrValue === 'function') {
       locationFields[field] = decoderOrValue(location.query[field]);
     } else {
-      staticFields[field] = decoderOrValue;
+      forwardedFields[field] = decoderOrValue;
     }
   }, {});
 
-  const stringyFields = JSON.stringify(locationFields);
-  const objFields = useMemo(() => JSON.parse(stringyFields), [stringyFields]);
+  const stringyForwardedFields = JSON.stringify(forwardedFields);
+  const stringyLocationFields = JSON.stringify(locationFields);
 
   return useMemo(
     () => ({
-      ...objFields,
-      ...staticFields,
+      ...JSON.parse(stringyForwardedFields),
+      ...JSON.parse(stringyLocationFields),
```

Comment:
i don't know if you even need JSON.parse since you already have `forwardedFields` and `locationFields`. the stringify does work to trigger the change detection though

Reply:
both of those are new objects each time, so each of them would trigger downstream updates whenever the hook runs. The stringify and useMemo are the main bits of this util to prevent those downstream renders.

Reply:
you can skip the json.parse and use the fields objects directly, keeping json.stringify

Reply:
but then eslint will complain again

Reply:
Removed the JSON.parse, so it's slightly faster whenever the location has changed by skipping that call.

But also, that was silently inserting `any` type into the system, which made things work, but i didn't like that it wasn't explicit. So now it's explicit.